allocator.rs

```diff
@@ -0,0 +1,197 @@
+//!
+//! When both configured, `"rust-allocator"` is preferred.
+//!
+//! The interface for the allocator is not a great fit for rust. In particular, rust always needs
+//! the layout of an allocation to deallocate it, and C interfaces don't usually provide this
+//! information. Luckily in the library we know in all cases how big the allocation was at the
+//! point where we deallocate it.
+use core::ffi::{c_int, c_void};
+
+type AllocFunc = unsafe extern "C" fn(*mut c_void, c_int, c_int) -> *mut c_void;
+type FreeFunc = unsafe extern "C" fn(*mut c_void, *mut c_void) -> ();
+
+pub(crate) enum Allocator {
+    #[cfg(feature = "rust-allocator")]
+    Rust,
+    #[cfg(feature = "c-allocator")]
+    C,
+    Custom {
+        allocate: AllocFunc,
+        deallocate: FreeFunc,
+        opaque: *mut c_void,
+    },
+}
+
+impl Allocator {
+    #[allow(unreachable_code)]
+    pub(crate) const DEFAULT: Option<Self> = 'blk: {
+        #[cfg(feature = "rust-allocator")]
+        break 'blk Some(Self::Rust);
+
+        #[cfg(feature = "c-allocator")]
+        break 'blk Some(Self::C);
+
+        None
+    };
+
+    pub(crate) fn from_bz_stream(strm: &crate::bz_stream) -> Option<Self> {
+        let bzalloc = strm.bzalloc?;
+        let bzfree = strm.bzfree?;
+
+        #[cfg(feature = "rust-allocator")]
+        if (bzalloc, bzfree) == rust_allocator::ALLOCATOR {
+            return Some(Self::Rust);
+        }
+
+        #[cfg(feature = "c-allocator")]
+        if (bzalloc, bzfree) == c_allocator::ALLOCATOR {
+            return Some(Self::C);
+        }
+
+        Some(Self::custom(bzalloc, bzfree, strm.opaque))
+    }
+
+    /// # Safety
+    ///
+    /// - `allocate` and `opaque` must form a valid allocator, meaning `allocate` returns either
+    ///     * a `NULL` pointer
+    ///     * a valid pointer to an allocation of `len * size_of::<T>()` bytes aligned to at least `align_of::<usize>()`
+    /// - `deallocate` frees memory allocated by `allocate`
+    pub(crate) fn custom(allocate: AllocFunc, deallocate: FreeFunc, opaque: *mut c_void) -> Self {
+        Self::Custom {
+            allocate,
+            deallocate,
+            opaque,
+        }
+    }
+
+    pub(crate) fn function_pointers(&self) -> (AllocFunc, FreeFunc) {
+        match self {
+            #[cfg(feature = "rust-allocator")]
+            Allocator::Rust => rust_allocator::ALLOCATOR,
+            #[cfg(feature = "c-allocator")]
+            Allocator::C => c_allocator::ALLOCATOR,
+            Allocator::Custom {
+                allocate,
+                deallocate,
+                ..
+            } => (*allocate, *deallocate),
+        }
+    }
+}
+
+#[cfg(feature = "c-allocator")]
+#[allow(non_camel_case_types)]
+type c_size_t = usize;
+
+#[cfg(feature = "c-allocator")]
+extern "C" {
+    fn malloc(size: c_size_t) -> *mut c_void;
+    fn calloc(nitems: c_size_t, size: c_size_t) -> *mut c_void;
+    fn free(p: *mut c_void);
```

Comment:
This should use the libc crate instead.

Reply:
is there a specific reason for that?

My plan was to add a `"stdio"` flag, which mirrors [`BZ_NO_STDIO`](https://www.sourceware.org/bzip2/manual/manual.html#stdio-bye). Then we could make the libc dependency conditional meaning it won't need to be compiled to use the low-level API.

Reply:
These functions are defined in the C library. Using the libc crate ensures that we use the right function signatures on all systems, even those that use different function signatures if any such systems exist. Also in case of a no_std build, we need the libc crate to tell the linker to link against the system libc for these definitions as rustc tells the linker to not link against any system libraries by default. And in the future if we ever get raw-dylib support for ELF and Mach-O systems, the libc crate will handle getting the right dylib name (eg macOS calls it `libSystem.dylib` rather than `libc.dylib`. `-lc` only works because the `libc.tbd` import library redirects to `libSystem.tbd`, but raw-dylib works without externally provided import libraries. And on Windows libc is called `msvcrt.dll` or `ucrt.dll` depending on which Windows target you use and on MSVC targets `-lc` doesn't work at all.) and symbol versions (eg `malloc@GLIBC_2.2.5` on glibc) for us.

Reply:
ok, fair enough, let's use libc for the C allocator then.

btw, `"stdio"` is not quite the same as `no_std`: the idea is that you could configure neither the rust nor c allocator (so you'd have to configure bzalloc and bzfree in the bz_stream struct). I think in that case we don't explicitly use any libc symbols.